application/src/main/java/org/openhds/mobile/provider/DatabaseAdapter.java

```diff
@@ -255,6 +281,91 @@ public void updateCompleteStatus(long id, boolean completed) {
 		updateFormSubmission(id, cv);
 	}
 
+	//add recent forms paths
+	public long addHierarchyPath(String hierarchyPath, String filePath) {
+		long id = -1;
+		SQLiteDatabase db = dbHelper.getWritableDatabase();
+		db.beginTransaction();
+		try {
+			ContentValues cv = new ContentValues();
+			cv.put(KEY_HIERARCHY_PATH, hierarchyPath);
+			cv.put(KEY_RECENT_FORM_PATH, filePath);
+			id = db.insertOrThrow(RECENT_FORM_TABLE, null, cv);
+			db.setTransactionSuccessful();
+		} finally {
+			db.endTransaction();
+		}
+		db.close();
+		return id;
+	}
+
+	//Finds recentForms by current hierarchy
+	public Collection findFormByPath(String hierarchyPath) {
+		Set<String> formpaths = new HashSet<>();
+		SQLiteDatabase db = dbHelper.getReadableDatabase();
+		Cursor cursor = null;
+		try {
+
+			cursor = db.query(RECENT_FORM_TABLE, new String[]{KEY_RECENT_FORM_PATH},
+					KEY_HIERARCHY_PATH + "= ?", new String[]{hierarchyPath}, null, null, null);
+
+			if (cursor == null) {
+				return null;
+			}
+
+			while (cursor.moveToNext()) {
+				String formPath;
+				formPath = cursor.getString(cursor.getColumnIndex(KEY_RECENT_FORM_PATH));
+				formpaths.add(formPath);
+			}
+			cursor.close();
+
+		} catch (Exception e) {
+			Log.w("findUserByUsername", e.getMessage());
+		}
+		return formpaths;
+	}
+
+
+	//Find all recent forms paths
+	public Collection findAllFormPaths () {
+
+		Set<String> allformpaths = new HashSet<>();
+		SQLiteDatabase db = dbHelper.getReadableDatabase();
+		Cursor cursor = null;
+		try {
+
+			cursor = db.query(RECENT_FORM_TABLE, new String[]{ KEY_RECENT_FORM_PATH },
+					null, null, null, null, null);
+
+			if (cursor == null) {
+				return null;
+			}
+
+			while (cursor.moveToNext()) {
+				String allformPath;
+				allformPath = cursor.getString(cursor.getColumnIndex(KEY_RECENT_FORM_PATH));
+				allformpaths.add(allformPath);
+			}
+			cursor.close();
+
+		} catch (Exception e) {
+			Log.w("findUserByUsername", e.getMessage());
+		}
+		return allformpaths;
+	}
+//delete submitted formsPaths
+	public void deleteSubmitForms(String sentFilepath) {
+		SQLiteDatabase db = dbHelper.getWritableDatabase();
+		db.beginTransaction();
+		db.delete(RECENT_FORM_TABLE, KEY_RECENT_FORM_PATH + " = ?", new String[] { "" + sentFilepath });
+		db.setTransactionSuccessful();
+		db.endTransaction();
+		db.close();
```

Comment:
I think we talked about this before, but didn't finish the conversation. Any place where we open resources like this, we should guarantee that we clean them up whether an exception occurs or not. Usually, this is done in a try/finally block.